### PR TITLE
moved caching from ocf/kubernetes#34 and reference XDG_CACHE_HOME

### PR DESCRIPTION
**Pull request by nelson-lojo**
Copied 'caching' solution from ocf/kubernetes#34 and now reference $XDG_CACHE_HOME instead of always building in `~/.cache`

Potential improvements:
- [x] check if some environmental variable is set that should override `$XDG_CACHE_HOME`
- [ ] delete old archives if more than 2 versions old?

### CODE REVIEW

File: transpire/helpers/helm.py [hunk 1]
```diff
@@ -95,6 +95,35 @@ def build_chart(
     if shutil.which("helm") is None:
         raise RuntimeError("You must install Helm to use this script.")
 
+    # make sure the current user has a place for us to cache stuff
+    home_cache = os.environ.get('XDG_CACHE_HOME', None)
+    if not home_cache:
+        raise EnvironmentError("Could not find an XDG_CACHE_HOME environment variable")
+    cache_dir = os.path.join(home_cache, 'ocf_helm')
```
Comment (nikhiljha) [inline transpire/helpers/helm.py]: We should rename this to `transpire` instead of `ocf_helm`.

File: transpire/helpers/helm.py [hunk 2]
```diff
@@ -95,6 +95,35 @@ def build_chart(
     if shutil.which("helm") is None:
         raise RuntimeError("You must install Helm to use this script.")
 
+    # make sure the current user has a place for us to cache stuff
+    home_cache = os.environ.get('XDG_CACHE_HOME', None)
+    if not home_cache:
+        raise EnvironmentError("Could not find an XDG_CACHE_HOME environment variable")
```
Comment (nikhiljha) [inline transpire/helpers/helm.py]: We definitely don't want to fail if `$XDG_CACHE_HOME` isn't set. In this case we probably want to default to `$HOME/.cache/transpire`.

File: transpire/helpers/helm.py [hunk 3]
```diff
@@ -95,6 +95,35 @@ def build_chart(
     if shutil.which("helm") is None:
         raise RuntimeError("You must install Helm to use this script.")
 
+    # make sure the current user has a place for us to cache stuff
+    home_cache = os.environ.get('XDG_CACHE_HOME', None)
```
Comment (nikhiljha) [inline transpire/helpers/helm.py]: I think the way we want this to work is...

1. If `TRANSPIRE_CACHE_DIR` is set, use `$TRANSPIRE_CACHE_DIR/charts` as the cache directory.
2. Failing that, if `XDG_CACHE_HOME` is set, use `$XDG_CACHE_HOME/transpire/charts` as the cache directory.
3. Failing that, if `HOME` is set, use `$HOME/.cache/transpire/charts` as the cache directory.
4. Failing that, use `/tmp/transpire/charts` as the cache directory.

Having this fail seems like an annoyance when before it worked in all cases. Also, this way we can support running transpire in CI with a persistent cache!

File: transpire/helpers/helm.py [hunk 4]
```diff
@@ -95,6 +95,35 @@ def build_chart(
...
+    if not chart_exists:
+        pull_args = [
+            "helm",
+            "pull",
+            "-d", cache_dir,
+            "--version", version,
+            "--repo", repo_url,
+            chart_name
+        ]
+        should_be_empty = run(
```
Comment (nikhiljha) [inline transpire/helpers/helm.py]: Did you mean to check this to see if it's empty, and error if it is?

**Review by nikhiljha — CHANGES_REQUESTED** [review]
Looks good! A few small changes and we can merge this in. Thank you for your contribution :)

File: transpire/helpers/helm.py (hunk 4 above)
Comment (nelson-lojo) [reply transpire/helpers/helm.py]: Yeah. I just added it.

File: transpire/helpers/helm.py (hunk 3 above)
Comment (nelson-lojo) [reply transpire/helpers/helm.py]: That makes a lot of sense. I figured that we'd want to pick another cache location for transpire at some point without affecting `$XDG_CACHE_HOME`, but I wasn't sure how to go about it.

File: transpire/helpers/helm.py (hunk 1 above)
Comment (nelson-lojo) [reply transpire/helpers/helm.py]: done!

File: transpire/helpers/helm.py [hunk 5]
```diff
@@ -95,6 +96,45 @@ def build_chart(
...
+        cache_dir = os.path.join(cache_dir, "transpire")
+    elif (cache_dir := os.environ.get("HOME", None)) is not None:
+        cache_dir = os.path.join(cache_dir, ".cache", "transpire")
+    else:
+        cache_dir = os.path.join("/tmp", "transpire")
+
+    if not os.path.isdir(cache_dir):
+        os.mkdir(cache_dir)
+
+    chart_dir: str = os.path.join(cache_dir, "charts")
```
Comment (nikhiljha) [inline transpire/helpers/helm.py:111]: Last thing: make sure this path is created!

**Review by nikhiljha — CHANGES_REQUESTED** [review]

File: transpire/helpers/helm.py [hunk 6]
```diff
@@ -117,7 +155,7 @@ def build_chart(
         f"ocf-{name}",
         "--api-versions",
         ", ".join(capabilities),
-        chart_name,
+        archive_name,
```
Comment (nikhiljha) [inline transpire/helpers/helm.py]: I think you need to give it the full path to the archive, or it won't be able to find it.

**Review by nikhiljha — CHANGES_REQUESTED** [review]

**Review by nikhiljha — APPROVED** [review]
Made the changes myself because they were small :)

**Pull request merged**